### PR TITLE
implement "nearly division less" algorithm for rand(a:b)

### LINKED ISSUE #20582: random sampling from an (abstract)array is slow
It is three times slower than this alternative implementation
```julia
julia> myrand(v) = (i = ceil(Int,rand()*length(v));  v[i])

julia> @benchmark rand(1:100)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     36.048 ns (0.00% GC)
  median time:      36.996 ns (0.00% GC)
  mean time:        37.104 ns (0.00% GC)
  maximum time:     66.210 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     993
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark myrand(1:100)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     11.824 ns (0.00% GC)
  median time:      12.332 ns (0.00% GC)
  mean time:        12.286 ns (0.00% GC)
  maximum time:     35.627 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

### PR DESCRIPTION
**Pull request by rfourquet**
Cf. https://arxiv.org/abs/1805.10941.
Closes #20582, #29004.

This is still about 30 percent slower than the "biased" solution (e.g. #28987), but as fast as the "best case" of the current version (when the range length is a power of 2). So this is strictly better, and should become the default version in a future (major) release. How do we make this available  in the meantime? `rand(Random.fast(a:b))` is what the PR currently implements.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Can't the behavior of `rand(a:b)` just be changed to use this without breaking the API?

**Comment by KristofferC** [comment]
Presumably, it changes what random numbers are generated.

**Comment by tkluck** [comment]
Arguably, it's only breaking API for those cases where the code uses `seed!`. So one way of preserving the API is as follows: in the new version of Julia, we set a default seed to something like `seed!(FasterRandom(<some seed>))`. If the user uses `seed!(::Integer)`, that's our cue to use the slower (but api-stable) random number generator from that point onwards.

**Comment by GregPlowman** [comment]
Just saw a reference to this on [Discourse](https://discourse.julialang.org/t/effective-simulation-of-putting-n-1-balls-in-n-boxes-uniformly-at-random/23774/12).
Could it be implemented now as `Future.rand()`?

**Comment by stevengj** [comment]
See also the discussion in #30494 — I'm skeptical that we should guarantee that the exact sequence of random number for a given seed is a stable part of the API.

**Comment by rfourquet** [comment]
> See also the discussion in #30494 — I'm skeptical that we should guarantee that the exact sequence of random number for a given seed is a stable part of the API.

I agree that that PR set a precedent, i.e. implicitly a rule that changing the stream of generated number is considered as "minor change". Should this be triaged to confirm?

Note though that we use a lot of `Random.seed!(123)` in tests, so changing the streams in a minor release means that some people would have to update their tests. Also, a change of streams between 0.6 and 0.7 was sufficiantly annoying to a user that he made a compatibility package to reproduce the 0.6 streams on 0.7+.

PS: I rushed a lot of PRs before the release of Julia 1 only because I was under the belief that we would guarantee the reproducibility of streams, I'm pretty sure this idea was in the air somehow... but it was for the best I guess!

File: stdlib/Random/src/generation.jl [hunk 1]
```diff
@@ -166,13 +166,15 @@ end
...
-#     consumption of entropy bits
+# there are three implemented samplers for unit ranges, the two first of which
+# assume that Float64 (i.e. 52 random bits) is the native type for the RNG:
+# 1) "Fast", which is most efficient when the underlying RNG produces rand(Float64)
+#    "fast enough". The tradeoff is faster creation of the sampler, but more
+#    consumption of entropy bits
 # 2) "Default" which tries to use as few entropy bits as possible, at the cost of a
 #    a bigger upfront price associated with the creation of the sampler
+# 3) "Nearly Division Less", which is the fastest algorithm for types of size
+#    up to 64 bits. This will be the default in a future realease.
```
Comment (halleysfifthinc) [inline stdlib/Random/src/generation.jl]:
```suggestion
#    up to 64 bits. This will be the default in a future release.
```

**Comment by stevengj** [comment]
People generally use Random.seed in tests not because they care about the exact sequence of random numbers but because they don’t want tests to fail in rare occasions because something extremely important improbable occurred (e.g. a random matrix happens to be nearly singular). By the same token, however, the same tests are extremely unlikely to fail if we change the pseudorandom stream.

**Comment by rfourquet** [comment]
Yes of course you are right, I somehow confused this use case with uses of `seed!` in `Random` tests.

**Comment by rfourquet** [comment]
While updating to make this algorithm (NDL) the default, I realised that we have a test garanteeing that e.g. `rand(1:9)` will give the same result regardless of the architecture (i.e. regardless of the value of `Int`). 

Unfortunately here this means that if we want to maintain this property, there is a (small) cost to pay. I may have made a mistake of course, but this is what I currently understand. Let's call "NDL Unified" the modification of `rand` in this PR that respect the garantee mentioned above, by adding this line at the beginning, to call the 32 bit version when the range length fits in 32 bits:
```
    if U === UInt64 && !iszero(sp.s) && sp.s <= typemax(UInt32)
        return rand(rng, SamplerRangeNDL(sp.a, sp.s % UInt32))
    end
```
Here are the observations (see graph below) for `MersenneTwister`, where input range is `UnitRange{Int64}` except for NDL 32 bits: 
- the default for `MersenneTwister` is `SamplerRangeFast`, whose efficiency decreases for a length from `2^n` to `2^(n+1)-1`, and is good again at `2^(n+1)`. There is a pic at `2^52` because `MersenneTwister` generates 52 bits natively (`SamplerRangeFast` is somewhat optimized for that)
- NDL (32 or 64 bits depending on input) follows the best case of `SamplerRangeFast`, with efficiency decreasing when approaching full capacity (2^64 here)
- NDL unified: on the `[1:2^32]` interval, the performance is not so attractive compared to the status quo (expect it's more predictable, no "pics"), in particular when approaching "full capacity" at `2^32`. I don't really understand why it becomes basically as efficient as NDL after 2^32
- NDL 32 bits, which corresponds to e.g. `rand(UInt32(1):UInt32(9))`: this is slightly better than NDL, and seems to reflect the difference between `rand(UInt32)` and `rand(UInt64)` involved in the computation (maybe about 1ns difference). 
- Another solution, NDL64 (opposite solution to NDL unified): not plotted, but should look like NDL, where even 32 bit ranges call the 64 bit version, unconditionally. The problem is that for RNGs where `rand(UInt64)` can be twice as expensive as `rand(UInt32)`, we really loose efficiency for small ranges, i.e. NDL32 (which would be more advantageous for this RNG than in this graph) would be unavailable.
- `SamplerRangeInt`: the default for non-`MersenneTwister`

compareSamplers

So, I my favorite solution is to drop the guarantee that `rand(1:9)` will give the result whether `Int` is 32 or 64 bits, as it gives access to the best performance, and it allows to specifically ask for the 32 version (by converting a range to 32 bits types) if deemed more favorable. If not acceptable, I would otherwise choose NDL64. What do you think?

**Comment by chethega** [comment]
This goes into the next release and won't be backported, right?

Then I am all for dropping the guarantees.

However, 
```
julia> s=Random.SamplerRangeNDL(UInt128(1):UInt128(2));
julia> @btime rand($s);
  477.703 ns (11 allocations: 186 bytes)

julia> s2=UInt128(1):UInt128(2);
julia> @btime rand($s2);
  25.301 ns (0 allocations: 0 bytes)
```

**Comment by rfourquet** [comment]
> This goes into the next release and won't be backported, right?

It would be great that it goes in the next release, but for sure won't be backported indeed!

> However, [...]

Yeah, there is a `widen` operation involved (which gives `BigInt` for `UInt128`), so this will be enabled only for types upto 64 bits. When you define `widen(::Type{UInt128}) = UInt256`, then NDL seems to still be faster than the status quo.
Also, it would be interesting to see if NDL can be extended efficiently to `BigInt`.

**Comment by GregPlowman** [comment]
Here's what seems like a minority voice (squeak, squeak).

When developing some simulations, I like to be able to compare results with previous runs. I guess this is a sort of unit test. I might experiment with algorithmic changes, performance enhancements etc. and then re-run the sim and check whether I'm still getting the same results.

In fact, I just upgraded from Julia v0.6 through v0.7, v1.0 and v1.1, and wanted to compare the new sim results to the old v0.6 ones. But I needed to implement the same custom `rand(1:n)` in both v0.6 and v0.7 so that the sequence was the same in both versions.

I understand the tension between progress and being held back by guarantees, especially when it is deemed the guarantee is of little benefit &/or the progress is of significant benefit.

However, I do want to point out the difference between random sequence guarantees across Julia versions, and across architectures within the same Julia version. I run my simulations across a network cluster and check the results against previous runs. In my case, all machines are 64-bit, but this would break in mixed-architecture clusters.

I don't like that the random sequence guarantee is broken across versions, but I'm happy to sacrifice that for progress. I feel less comfortable if the guarantee is broken across architectures on the same Julia version.

**Comment by rfourquet** [comment]
I appreciate the usefulness of compatibility between architectures, however it's already not compatible for `rand(Int)`, but more importantly it seems to me that using Julia on 32-bits architectures is really marginal. So everyone having to pay a price for something quite rare (unless my impression is mistaken) doesn't look like a good balance to me.

I would favor having a opt-in way of getting reproducibility. One of them would be to simply create the range with `Int32` values, e.g. `Int32(1):Int32(9)`. Another would be to have an RNG (e.g. a wrapper of `MersenneTwister`) which offers some guarantees. If this RNG was "frozen", we could even imagine being able to guarantee stream reproducibility across some Julia versions for some types/ranges.
Just my 2c.

**Comment by rfourquet** [comment]
Also, I just read Numpy's policy on RNG https://www.numpy.org/neps/nep-0019-rng-policy.html, this may be relevant to the discussion.

**Comment by chethega** [comment]
Can we put up some docs for an official reproducibility + rng-stream policy? Is there an issue/PR that discusses that?

As far as I understood, (1) we have no reproducibility between 32 bit and 64 bit anyway, (2) new releases may break RNG streams [32208](https://github.com/JuliaLang/julia/pull/32208), [30617](https://github.com/JuliaLang/julia/pull/30617), (3) we attempt to avoid breaking RNG streams on bugfix / backports.

Do we attempt reproducibility between architectures (e.g. aarch64 vs x86_64 vs power)?

Since we break RNG streams on releases, it would be almost sensible to codify this by having seeding of the RNG incorporate the version (like `1.2`, `1.3`, etc) and potentially word size and potentially architecture. That way, everybody knows that a new release will change all random streams, and randomized testsets will need to be regenerated. Nobody will then incorrectly rely on cross-version reproducibility of random streams. 

The worst case is random streams being almost reproducible, which is what we have now (and what numpy appears to be doing). 

If we codify this dependence of RNG streams, then we also lay the groundwork for replacing mersenne twister with alternatives that are faster (e.g. xoroshiro) or better (i.e. cryptographic like aes-ctr with hardware acceleration).

**Comment by rfourquet** [comment]
> Can we put up some docs for an official reproducibility + rng-stream policy? Is there an issue/PR that discusses that?

The [comment](https://github.com/JuliaLang/julia/pull/30494#issuecomment-449741435) from which I grabbed the link I posted above was just requesting that, and there was a small discussion yesterday on slack (triage channel) on this topic too. There is no open issue AFAIK though.

I find your idea of "codifying" in each release the breaking behavior in the seeding process quite interesting!

**Comment by stevengj** [comment]
Official docs on reproducibility of random streams was merged in #33350.

**Comment by rfourquet** [comment]
> Official docs on reproducibility of random streams was merged in #33350.

This is great, but does not explicitly cover the question (which needs to be resolved here): must `rand(1:9)` generate the same stream of numbers on 32 and 64 bits architectures? I believe we never officially guaranteed that, but an explicit effort was made to have this behavior, so an explicit decision must be made to break it.

**Comment by rfourquet** [comment]
> I have also for some time wanted to hash seed values with a cryptographic hash like SHA1 or SHA256 before passing that seed value on to the underlying RNG for actual seeding

Sounds like a great idea, if no-one beats me to it, I will try to implement a POC when I have some time.

**Comment by stevengj** [comment]
> This is great, but does not explicitly cover the question (which needs to be resolved here): must `rand(1:9)` generate the same stream of numbers on 32 and 64 bits architectures?

I think this is a good property to have.

Suppose you have a stochastic program that is working on a 64-bit machine but failing on 32-bit, and you are trying to isolate the source of the failure.   Not being able to generate the same random number stream on both architectures with the same Julia versions seems like it would make debugging a lot harder.

**Comment by rfourquet** [comment]
> I think this is a good property to have.

Certainly, but in the present case it's at the cost of performance, and there is a work-around (use `Int32(a):Int32(b)`).
@mschauer would you share your opinion on this?

**Comment by mschauer** [comment]
I overlooked this issue. Let me come back.

**Comment by StefanKarpinski** [comment]
I think we can generally optimize for 64-bit systems at the expense of 32-bit ones. If people need really high performance, they're generally not working on 32-bit computers these days.

**Comment by rfourquet** [comment]
> I think we can generally optimize for 64-bit systems at the expense of 32-bit ones

I fully agree. But even then, we have a decision to make: do we keep "stream compatibility" between 32 and 64 architectures (i.e. `Random.seed!(0); rand(1:9)` gives the same output whatever `Int` is)?
* if so, we have the choice between "NDL-unified" and "NDL64" as described [above](https://github.com/JuliaLang/julia/pull/29240#issuecomment-497363751). If we favor "optimize for 64-bit systems", then we choose "NDL64", which is a pessimization for 32-bit systems.
* otherwise, each architecture gets its native algorithm ("NDL").

I would vote for the latter option.

Note that with other RNGs, 64 bits generation might be the "native output" even on 32-bits systems, in which case it's possible to have "sream compatibility" for free.

Off-Topic: since a couple of days, I was thinking to implement passing seeds to a crypto hash function, and I [realize now](https://github.com/JuliaLang/julia/pull/29240#issuecomment-534612829) it was not at all an original thought of mine :see_no_evil: !

**Comment by mschauer** [comment]
I feel that I was getting mixed signals, so I'd like to ask: What is our take, does architectures with small integers become less important for us (because being outdated on the personal computer) or more important (because of developments in lower precision cloud computing)?

**Comment by StefanKarpinski** [comment]
I'm unfamiliar with a trend towards lower precision cloud computing. Are you referring to GPUs and other hardware using Float16 and BFloat16?

**Comment by rfourquet** [comment]
I labeled for triage, a decision is needed here.

**Comment by KristofferC** [comment]
Triage is ok but let's open issues on packages that start failing their tests due to assuming a specific random number stream for this call.

@nanosoldier `runtests(ALL; vs =":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/29240#issuecomment-611709151) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/3b9f05ce1b40a3205f02bef257dbbb17a2eb3e5a/pkgeval/by_hash/4f58f53_vs_8a5b323/report.md). cc @maleadt

**Comment by rfourquet** [comment]
Thanks @KristofferC for bringing this issue up during the triage call.

I now made NDL the default algorithm for bitstypes up to 64 bits (NDL doesn't work well beyond, unless we have `Int256` etc., as widening is used). And for `[U]Int128`, I made `SamplerRangeFast` the default (it was the case only for `MersenneTwister` till now), as I think most used RNGs tend to be fast (in which case `SamplerRangeFast` is... well, faster). It's still possible for slower RNGs to opt-in for `SamplerRangeInt`. 

So the notable change is, as mentioned above, the fact that `rand(Int32(1):Int32(3))` and `rand(Int64(1):Int64(3))` don't produce the same streams anymore, given a seed. It's still possible to get the same streams on both architectures either by using an explicitly typed range (e.g. `Int64(1):3`), or by using explicitly another sampler, e.g. `SamplerRangeInt`.

**Comment by rfourquet** [comment]
I reviewed again the paper (https://arxiv.org/abs/1805.10941) and the implementation, but it would be good if someone can make a review! (at least of the code, corresponding to Algorithm 5 in the paper, page 6).

**Comment by mschauer** [comment]
@rfourquet Should we meet on chat sometimes next week to discuss Algorithm 5?

**Comment by rfourquet** [comment]
> Should we meet on chat sometimes next week to discuss Algorithm 5?

Sure! Ping me when you are ready. If you want to discuss the theory behind the algo, maybe tell me one day in advance so that I have time to refresh my memory :)

**Review by mschauer — APPROVED** [review]
Discussed the implementation of algorithm 5 with Rafael via chat.

**Comment by rfourquet** [comment]
Thanks @mschauer ! I added a comment for `mod(-s, s)` being equivalent to `2^L - s mod s` in the paper.
If no-one objects, I will then merge this week to get this in for 1.5.

**Comment by rfourquet** [comment]
@nanosoldier `runtests(ALL; vs =":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/29240#issuecomment-621756660) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/ab6676206b210325500b4f4619fa711f2d7429d2/pkgeval/by_hash/52c2272_vs_47c55db/report.md). cc @maleadt

**Comment by rfourquet** [comment]
Looks like a good time to merge this... does someone like to press the button?

**Comment by StefanKarpinski** [comment]
I love pressing buttons!

**Comment by rfourquet** [comment]
Thanks! it helps with feeling that the responsibility is shared ;-)

**Pull request merged**